<a href="https://colab.research.google.com/github/Mahal09/Python/blob/main/Week2_Numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 3]


In [ ]:
np_arr_str = np.array (arr_str)
np_arr_num = np.array(arr_num)

print(np_arr_str)
print(type(np_arr_num))

['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
<class 'numpy.ndarray'>


In [ ]:
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


In [ ]:
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print( matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [ ]:
print(type( matrix))

<class 'numpy.ndarray'>


In [ ]:
arr2  = np.arange(start = 0, stop = 10)
# 10 will be excluded from the output
print(arr2)

# or

arr2  = np.arange(0,10)
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [ ]:
arr3 = np.arange(0,20,5)
print(arr3)

[ 0  5 10 15]


In [ ]:
arr3 = np.arange(0,20,5)
arr3

array([ 0,  5, 10, 15])

In [ ]:
arr3  = np.arange(start = 0, stop = 20, step = 5)
arr3

array([ 0,  5, 10, 15])

In [ ]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
type(matrix2)
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

How are these values getting generated?

The step size or the difference between each element will be decided by the following formula:

(stop - start) / (total elements - 1)

So, in this case: (5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on

In [ ]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,11)
matrix3

array([10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20.])

20-10/11-1 =10/10

In [ ]:
matrix4 = np.zeros([3,5])
matrix4

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [ ]:
matrix4 = np.zeros([3,15])
matrix4

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [ ]:
matrix5 = np.ones([3,7])
matrix5

array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]])

In [ ]:
matrix5 = np.ones([2,2])
matrix5

array([[1., 1.],
       [1., 1.]])

In [ ]:
# prompt: what is the use of Using np.eye()

import numpy as np

# Create a 3x3 identity matrix
identity_matrix = np.eye(3)
print(identity_matrix)

# Explanation:
# np.eye(n) creates an n x n identity matrix.
# An identity matrix has 1s on the main diagonal and 0s everywhere else.
# It's useful for linear algebra operations and matrix transformations.


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
matrix6 = np.eye(4)
print(matrix6)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


In [ ]:
vec = np.zeros([2,3])
print(vec)

[[0. 0. 0.]
 [0. 0. 0.]]


In [ ]:
vec = np.linspace(10,20,3)
vec1 = np.linspace(20,40,2)
print(vec1)

print(vec+vec1)

[20. 40.]


ValueError: operands could not be broadcast together with shapes (3,) (2,) 

In [ ]:
arr4 = np.arange(0,10)
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
arr4_reshaped = arr4.reshape(2,5)
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [ ]:
arr5 = np.array([10,2,6])
np.exp(arr5)

array([2.20264658e+04, 7.38905610e+00, 4.03428793e+02])

In [ ]:
print('Sine function',np.sin(30))

print('Cosine function',np.cos(30))
print('Tan Function',np.tan(60))


Sine function -0.9880316240928618
Cosine function 0.15425144988758405
Tan Function 0.320040389379563


In [ ]:
# prompt: why everything in np comes with more floating point numbers

# NumPy often uses floating-point numbers because it's the default for many mathematical operations.
# This is due to the nature of numerical computing and the way NumPy is designed to handle various mathematical functions and operations efficiently.

# Here's why:

# 1.  Mathematical Functions: Many mathematical functions in NumPy (like trigonometric functions, logarithms, etc.) are defined for floating-point numbers. Using floating-point numbers allows for a wider range of values and more precise calculations.

# 2.  Compatibility with Other Libraries:  Floating-point numbers are the standard for scientific computing and data analysis.  This ensures compatibility with other libraries and tools that rely on floating-point numbers.

# 3.  Efficiency: Floating-point operations are highly optimized in hardware and software, making computations faster.

# 4.  Precision: While integers are suitable for discrete values, floating-point numbers offer greater precision for representing real-world data and calculations involving decimals.


# To illustrate this, let's look at an example:
import numpy as np

arr = np.array([1, 2, 3])
print(arr.dtype)  # Output: int64

arr = np.array([1, 2, 3], dtype=float)
print(arr.dtype)  # Output: float64

# As you can see, by default, NumPy might infer the data type as int64. However, you can explicitly specify the data type as float64 if needed.


In [ ]:
# prompt: np.exp(2) , how and why

import numpy as np

result = np.exp(2)
print(result)

# Explanation:
# np.exp(2) calculates the exponential value of 2.
# It means e raised to the power of 2, where 'e' is the base of the natural logarithm (approximately 2.71828).
# The result will be approximately 7.38905609893065.


In [ ]:
np.log(arr5)

array([2.30258509, 0.69314718, 1.79175947])

In [ ]:
np.log10(8)

0.9030899869919435

In [ ]:
np.log(2)

0.6931471805599453

In [ ]:
vec1 = np.arange(0,12)
print(vec1)
vec2 = vec1.reshape(2,6)
print(vec2)

[ 0  1  2  3  4  5  6  7  8  9 10 11]
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]


In [ ]:
vec = np.arange(0,5)
print(vec)
print(vec.reshape(1,5))

[0 1 2 3 4]
[[0 1 2 3 4]]


In [ ]:
l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)

[1, 2, 3, 4, 5, 6]


In [ ]:
arr7 = np.arange(1,6)
print('arr7:',arr7)

arr8 = np.arange(3,8)
print('arr8:',arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [ ]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!

Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


In [ ]:
matrix7 =np.arange(1,10).reshape(3,3)
print(matrix7)
matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
print('Addition transpose: \n',np.transpose(matrix7+matrix8))
print('addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition transpose: 
 [[ 2.  4.  7.]
 [ 2.  6.  8.]
 [ 3.  6. 10.]]
addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


<ipython-input-69-acded3ad6175>:5: RuntimeWarning: divide by zero encountered in divide
  print('Division: \n', matrix7/matrix8)


In [ ]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

In [ ]:
veec = np.arange(2,5,1)
print(1/veec)
print(veec**2)

[0.5        0.33333333 0.25      ]
[ 4  9 16]


In [ ]:
mat1 = np.arange(1,5).reshape(2,2)
mat2 = np.eye(2)
print(mat1/mat2)

[[ 1. inf]
 [inf  4.]]


<ipython-input-84-0e6adcba2ca7>:3: RuntimeWarning: divide by zero encountered in divide
  print(mat1/mat2)


In [ ]:
0.25*2

0.5

In [ ]:
0.5*2

1.0

In [ ]:
rand_mat = np.random.randn(5)
print(rand_mat)

[-0.18573995  0.2154751   0.23664706  0.46837086 -0.30108808]


In [ ]:
rand_mat2 = np.random.rand(5)
print(rand_mat2)

[0.22803719 0.98893504 0.68396906 0.7979226  0.11277643]


In [ ]:
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[ 0.1527305   1.10241297 -0.06482158 -0.72423698  1.32944484]
 [ 0.19751992 -0.54389682 -0.4049916  -2.23888652 -2.19089248]
 [-0.60097248  0.74918037 -0.93624118 -0.71758916 -0.0912532 ]
 [ 0.23619711 -0.13290724 -1.1624427  -0.43542412 -1.299046  ]
 [-0.63402732  1.07610857 -0.85721652 -2.68582467 -0.53857891]]


In [ ]:
print(np.mean(rand_mat2))
print(np.std(rand_mat2))

0.5623280613964007
0.33649424648743237


In [4]:
rand_mat3=np.random.randint(1,5,10)
print(rand_mat3)

[3 2 4 4 2 3 4 1 3 3]


In [5]:
rand_mat3=np.random.randint(1,10,[5,5])
print(rand_mat3)

[[3 8 2 7 5]
 [8 6 7 4 1]
 [9 5 9 6 6]
 [1 9 9 5 9]
 [7 4 8 7 1]]


In [ ]:
mat1 = np.arange(-9,0,1).reshape(3,3)
print(mat1)
print(np.min(mat1))
print(np.max(mat1))

[[-9 -8 -7]
 [-6 -5 -4]
 [-3 -2 -1]]
-9
-1


In [ ]:
rand_arr = np.random.randn(10)
print(rand_arr)

[-0.85963106 -0.18092958  0.67766152  1.1387774  -1.00989886 -1.11470844
 -1.88615618 -0.9415078   0.1923508   0.88548301]


In [ ]:
print(rand_arr[6])

0.9359749929087581


In [ ]:
print(rand_arr[4:9])

[-0.88759036  0.46493157  0.93597499 -0.67094174  1.4345913 ]


In [ ]:
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[-0.15694803  0.93597499 -0.59469591]


In [ ]:
print(rand_arr)

[ 0.51965182 -0.54848652  0.93328396 -0.15694803 -0.88759036  0.46493157
  0.93597499 -0.67094174  1.4345913  -0.59469591]


In [ ]:
rand_arr>0

array([ True, False,  True, False, False,  True,  True, False,  True,
       False])

In [ ]:
rand_arr < 0

array([False,  True, False,  True,  True, False, False,  True, False,
        True])

In [ ]:
# prompt: what is rand and randn

# np.random.rand(n) generates an array of n random numbers uniformly distributed between 0 and 1.
# Example:
rand_mat2 = np.random.rand(5)
print(rand_mat2)

# np.random.randn(n) generates an array of n random numbers from the standard normal distribution (mean 0, variance 1).
# Example:
rand_mat = np.random.randn(5)
print(rand_mat)


[0.77779466 0.33742638 0.88135179 0.87037203 0.45493473]
[ 1.24646011 -0.67627726  0.52810239  0.77669637  0.53859691]


In [ ]:
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[-0.61256465  3.01948589  1.62897345 -0.02786428 -0.64398957]
 [ 0.00701759  1.8255322  -0.25623488  1.39796888  2.0065455 ]
 [-1.54025889  0.25422903 -1.34589915  0.5359942  -0.74825582]
 [-0.77858219 -1.59906044  1.44702477  1.34960827  0.11078514]
 [ 0.72174153 -0.68999413  0.69589443 -0.3879187  -0.78007918]]


In [ ]:
rand_mat[1]

array([ 0.00701759,  1.8255322 , -0.25623488,  1.39796888,  2.0065455 ])

In [ ]:
print(rand_mat[1][2])

-0.25623488365068314


In [ ]:
print(rand_mat[0:2,1:3])

[[ 3.01948589  1.62897345]
 [ 1.8255322  -0.25623488]]


In [ ]:
print(rand_mat)

[[-0.61256465  3.01948589  1.62897345 -0.02786428 -0.64398957]
 [ 0.00701759  1.8255322  -0.25623488  1.39796888  2.0065455 ]
 [-1.54025889  0.25422903 -1.34589915  0.5359942  -0.74825582]
 [-0.77858219 -1.59906044  1.44702477  1.34960827  0.11078514]
 [ 0.72174153 -0.68999413  0.69589443 -0.3879187  -0.78007918]]


In [ ]:
print([rand_mat >0])

[array([[False,  True,  True, False, False],
       [ True,  True, False,  True,  True],
       [False,  True, False,  True, False],
       [False, False,  True,  True,  True],
       [ True, False,  True, False, False]])]


In [ ]:
print([rand_mat<0])

[array([[ True, False, False,  True,  True],
       [False, False,  True, False, False],
       [ True, False,  True, False,  True],
       [ True,  True, False, False, False],
       [False,  True, False,  True,  True]])]


In [ ]:
print(rand_mat[rand_mat>0])

[3.01948589 1.62897345 0.00701759 1.8255322  1.39796888 2.0065455
 0.25422903 0.5359942  1.44702477 1.34960827 0.11078514 0.72174153
 0.69589443]


In [ ]:
print(rand_arr)

[-0.85963106 -0.18092958  0.67766152  1.1387774  -1.00989886 -1.11470844
 -1.88615618 -0.9415078   0.1923508   0.88548301]


In [ ]:
rand_arr[5] =9
print(rand_arr)

[-0.85963106 -0.18092958  0.67766152  1.1387774  -1.00989886  9.
 -1.88615618 -0.9415078   0.1923508   0.88548301]


In [ ]:
rand_arr[2:4] = 5
print(rand_arr)

[-0.85963106 -0.18092958  5.          5.         -1.00989886  9.
 -1.88615618 -0.9415078   0.1923508   0.88548301]


In [ ]:
matrix = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(matrix[1][1])

5


In [6]:
print(rand_mat3)

[[3 8 2 7 5]
 [8 6 7 4 1]
 [9 5 9 6 6]
 [1 9 9 5 9]
 [7 4 8 7 1]]


# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0

In [7]:
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

Matrix before modification: 
 [[3 8 2 7 5]
 [8 6 7 4 1]
 [9 5 9 6 6]
 [1 9 9 5 9]
 [7 4 8 7 1]]
Matrix after modification: 
 [[3 8 2 7 5]
 [8 6 7 0 0]
 [9 5 9 0 0]
 [1 9 9 5 9]
 [7 4 8 7 1]]


In [8]:
matrix = np.arange(1,10).reshape(3,3)
matrix[0:2,1:2] = 0
print(matrix)

[[1 0 3]
 [4 0 6]
 [7 8 9]]
